pkg/common/oras/authprovider/alibabacloud/alibabacloudacrbasic.go

```diff
@@ -0,0 +1,198 @@
+
+	getAuthorizationTokenRequest := &cr20181201.GetAuthorizationTokenRequest{}
+	getAuthorizationTokenRequest.InstanceId = tea.String(d.instanceID)
+	runtime := &util.RuntimeOptions{}
+
+	tokenResponse, err := crClient.GetAuthorizationTokenWithOptions(getAuthorizationTokenRequest, runtime)
+	if err != nil {
+		return AcrAuthToken{}, fmt.Errorf("failed to init alibaba cloud acr client: %w", err)
+	}
+	d.acrAuthToken.AuthData[registry] = tokenResponse.Body
+
+	return d.acrAuthToken, nil
+}
+
+// Create returns an Alibaba CloudAcrBasicProvider
+func (s *AlibabaCloudAcrBasicProviderFactory) Create(authProviderConfig provider.AuthProviderConfig) (provider.AuthProvider, error) {
+	conf := alibabacloudAcrBasicAuthProviderConf{}
+	authProviderConfigBytes, err := json.Marshal(authProviderConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = json.Unmarshal(authProviderConfigBytes, &conf); err != nil {
+		return nil, fmt.Errorf("failed to parse auth provider configuration: %w", err)
+	}
+	//get ACR EE instance id
+	instanceID := conf.InstanceID
+	if instanceID == "" {
+		instanceID = os.Getenv("ALIBABA_CLOUD_ACR_INSTANCE_ID")
+		if instanceID == "" {
+			return nil, re.ErrorCodeEnvNotSet.WithComponentType(re.AuthProvider).WithDetail("no instance ID provided and ALIBABA_CLOUD_ACR_INSTANCE_ID environment variable is empty")
+		}
+	}
+
+	acrAuthToken := AcrAuthToken{}
+	acrAuthToken.AuthData = make(map[string]*cr20181201.GetAuthorizationTokenResponseBody)
+
+	return &alibabacloudAcrBasicAuthProvider{
+		acrAuthToken: acrAuthToken,
+		instanceID:   instanceID,
+		providerName: alibabacloudAcrAuthProviderName,
+	}, nil
+}
+
+// Enabled checks for non-empty AlibabaCloud RAM creds
+func (d *alibabacloudAcrBasicAuthProvider) Enabled(_ context.Context) bool {
+	if d.providerName == "" {
+		logrus.Error("basic Alibaba Cloud ACR providerName was empty")
+		return false
+	}
+
+	return true
+}
+
+// Provide returns the credentials for a specified artifact.
+// Uses AlibabaCloud RRSA to retrieve creds from RRSA credential chain
+func (d *alibabacloudAcrBasicAuthProvider) Provide(ctx context.Context, artifact string) (provider.AuthConfig, error) {
+	logrus.Debugf("artifact = %s", artifact)
+
+	if !d.Enabled(ctx) {
+		return provider.AuthConfig{}, fmt.Errorf("Alibaba Cloud RRSA basic auth provider is not properly enabled")
+	}
+
+	registry, err := provider.GetRegistryHostName(artifact)
+	if err != nil {
+		return provider.AuthConfig{}, errors.Wrapf(err, "could not get ACR registry from %s", artifact)
+	}
+
+	if !d.acrAuthToken.exists(registry) {
+		logrus.Debugf("acrAuthToken for %s does not exist", registry)
+		_, err = d.getAcrAuthToken(artifact)
+		if err != nil {
+			return provider.AuthConfig{}, errors.Wrapf(err, "could not get ACR auth token for %s", artifact)
+		}
+	}
+
+	// need to refresh AlibabaCloud ACR credentials
+	t := time.Now().Add(time.Minute * 5)
+	if t.After(d.acrAuthToken.Expiry(registry)) || time.Now().After(d.acrAuthToken.Expiry(registry)) {
```

Comment:
Seems the second condition check can be removed

Reply:
fixed